src/pool/pool_tbb.c

```diff
@@ -0,0 +1,238 @@
+    umf_result_t ret =
+        umfMemoryProviderAlloc(pool->mem_provider, *raw_bytes, 0, &resPtr);
+    if (ret != UMF_RESULT_SUCCESS) {
+        TLS_last_allocation_error = UMF_RESULT_ERROR_OUT_OF_HOST_MEMORY;
+        return NULL;
+    }
+
+    return resPtr;
+}
+
+static void tbb_raw_free_wrapper(intptr_t pool_id, void *ptr, size_t bytes) {
+    struct tbb_memory_pool *pool = (struct tbb_memory_pool *)pool_id;
+    umf_result_t ret = umfMemoryProviderFree(pool->mem_provider, ptr, bytes);
+    if (ret != UMF_RESULT_SUCCESS) {
+        fprintf(
+            stderr,
+            "Memory provider failed to free memory, addr = %p, size = %lu\n",
+            ptr, bytes);
+    }
+}
+
+static umf_result_t tbb_pool_initialize(umf_memory_provider_handle_t provider,
+                                        void *params, void **pool) {
+    (void)params; // unused
+
+    const size_t GRANULARITY = 2 * 1024 * 1024;
+    struct mem_pool_policy_s policy = {.pAlloc = tbb_raw_alloc_wrapper,
+                                       .pFree = tbb_raw_free_wrapper,
+                                       .granularity = GRANULARITY,
+                                       .version = 1,
+                                       .fixed_pool = false,
+                                       .keep_all_memory = false,
+                                       .reserved = 0};
+
+    pthread_once(&tbb_is_initialized, load_tbb_symbols);
+
+    struct tbb_memory_pool *pool_data = malloc(sizeof(struct tbb_memory_pool));
+    if (!pool_data) {
+        fprintf(stderr, "cannot allocate memory for metadata\n");
+        return UMF_RESULT_ERROR_OUT_OF_HOST_MEMORY;
+    }
+
+    pool_data->mem_provider = provider;
+    g_tbb_ops.pool_create_v1((intptr_t)pool_data, &policy,
+                             &(pool_data->tbb_pool));
+    *pool = (void *)pool_data;
+
+    return UMF_RESULT_SUCCESS;
+}
+
+static void tbb_pool_finalize(void *pool) {
+    pthread_once(&tbb_is_initialized, load_tbb_symbols);
+    struct tbb_memory_pool *pool_data = (struct tbb_memory_pool *)pool;
+    g_tbb_ops.pool_destroy(pool_data->tbb_pool);
+}
+
+static void *tbb_malloc(void *pool, size_t size) {
+    struct tbb_memory_pool *pool_data = (struct tbb_memory_pool *)pool;
+    TLS_last_allocation_error = UMF_RESULT_SUCCESS;
+    void *ptr = g_tbb_ops.pool_malloc(pool_data->tbb_pool, size);
+    if (ptr == NULL) {
+        if (TLS_last_allocation_error == UMF_RESULT_SUCCESS) {
+            TLS_last_allocation_error = UMF_RESULT_ERROR_UNKNOWN;
+        }
+        return NULL;
+    }
+    return ptr;
+}
+
+static void *tbb_calloc(void *pool, size_t num, size_t size) {
+    assert(pool);
+    size_t csize = num * size;
+    void *ptr = tbb_malloc(pool, csize);
+    if (ptr == NULL) {
+        // TLS_last_allocation_error is set by tbb_malloc()
+        return NULL;
+    }
+
+    memset(ptr, 0, csize); // TODO: device memory is not accessible by host
```

Comment:
I think this TODO is not needed here. I don't think we can use tbb pool with USM anyway (in TBB, metadata is stored alongside the data).

Reply:
Done